### PR TITLE
build: sign binary files in macos package

### PR DESCRIPTION
**Pull request by id**
Works on my machine™.

Need to setup secrets and try it out here.

### CODE REVIEW

File: scripts/macos-sign-binaries.sh [hunk 1]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
+
+# intended to run on MacOS only
+# signs all executable files in a given folder (as $1) with developer certificate
+
+# required variables:
+# APPLE_DEVELOPER_IDENTITY: "Developer ID Application: <company name> (<hex id>)"
+# APPLE_DEVELOPER_ID_BUNDLE: base64-encoded content of apple developer id certificate bundle in pksc12 format
+# APPLE_DEVELOPER_ID_BUNDLE_PASSWORD: password used when exporting the bundle
```
Comment (zmstone) [inline scripts/macos-sign-binaries.sh:9]: if this is to be a part of the CI,
should we add them to github secrents ? or should we run int in a private self-hosted runner ?

Comment (id) [reply scripts/macos-sign-binaries.sh:9]: github secrets should work

File: scripts/macos-sign-binaries.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+KEYCHAIN='emqx.keychain-db'
+KEYCHAIN_PASSWORD="$(openssl rand -base64 100)"
+
+security create-keychain -p "${KEYCHAIN_PASSWORD}" "${KEYCHAIN}"
+
+security import "${PKSC12_FILE}" -P "${APPLE_DEVELOPER_ID_BUNDLE_PASSWORD}" \
+         -t cert -f pkcs12 -k "${KEYCHAIN}" -T /usr/bin/codesign
+
+find "${1}" -perm -a=x -type f -not -name '*.so' -exec \
```
[the remark this answers is not in the document]
Comment (id) [reply scripts/macos-sign-binaries.sh:26]: yes, but no changes made to the scripts themselves

File: scripts/macos-sign-binaries.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+# add new keychain into the search path for codesign, otherwise the stuff does not work
+keychains=$(security list-keychains -d user)
+keychain_names=();
+for keychain in ${keychains}
+do
+  basename=$(basename "${keychain}")
+  keychain_name=${basename::${#basename}-4}
+  keychain_names+=("${keychain_name}")
+done
```
Comment (zmstone) [inline scripts/macos-sign-binaries.sh:37]: nit: following existing bash convention
1) `for` and `do` in the same line
2) width=4 indentation

```suggestion
for keychain in ${keychains}; do
    basename=$(basename "${keychain}")
    keychain_name=${basename::${#basename}-4}
    keychain_names+=("${keychain_name}")
done
```

File: scripts/macos-sign-binaries.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+security -v list-keychains -s "${keychain_names[@]}" "${KEYCHAIN}"
+
+set -x
```
Comment (zmstone) [inline scripts/macos-sign-binaries.sh:40]: wouldn't this expose the secret ?
```suggestion
set +x
```

File: scripts/macos-sign-binaries.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+REL_DIR="${1}"
```
Comment (zmstone) [inline scripts/macos-sign-binaries.sh:41]: move this to the top of the file.

File: .github/workflows/build_slim_packages.yaml [hunk 6]
```diff
@@ -150,30 +150,38 @@ jobs:
...
       run: |
         kerl update releases
         kerl build ${{ matrix.otp }}
         kerl install ${{ matrix.otp }} $HOME/.kerl/${{ matrix.otp }}
     - name: build ${{ matrix.profile }}
       env:
         AUTO_INSTALL_BUILD_DEPS: 1
+        APPLE_SIGN_BINARIES: 1
+        APPLE_ID: developers@emqx.io
+        APPLE_TEAM_ID: 26N6HYJLZA
```
Comment (qzhuyan) [inline .github/workflows/build_slim_packages.yaml:170]: is this a secret?

Comment (id) [reply .github/workflows/build_slim_packages.yaml:170]: It's visible in the binary, e.g.

```
codesign -d -v erts-12.2.1/bin/erl
Executable=/Users/id/Downloads/macos/emqx/erts-12.2.1/bin/erl
Identifier=erl
Format=Mach-O thin (x86_64)
CodeDirectory v=20500 size=591 flags=0x10000(runtime) hashes=13+2 location=embedded
Signature size=4668
Signed Time=12 Sep 2022 at 11:41:33
Info.plist=not bound
TeamIdentifier=26N6HYJLZA
Runtime Version=12.1.0
Sealed Resources=none
Internal requirements count=1 size=164
```

**Review by zmstone — APPROVED** [review]

**Pull request merged**